### PR TITLE
Add MySQL query metricset with lightweight module and SQL helper

### PR DESCRIPTION
**Pull request by sayden**
## What does this PR do?

This PR adds a SQL helper (at Metricbeat level) with logic to execute SQL queries and return their results. The module can use this abstraction layer to simply pass in the SQL driver and the queries they need to execute to later create Beats events.

This PR introduce:
* Major changes in `sql` module.
* Completely new SQL helper with refactored code from `sql` module.
* `query` Metricset in MySQL module which uses the SQL helper. This Metricset is to use it as a Lightweight "parent" module.
* `performance` metricset using the `query` metricset as Lightweight parent module. Use it as a guidance about how to make it work.
* ~~`overview` metricset with a bunch of key metrics, also using Lightweight modules. This is the metricset to remove of this PR.~~ Removed
* Rework of the SQL module to use the SQL helper too. Basically the code from `sql` module was moved to the `sql` helper.

## Why is it important?
The way the `sql` was working was a bit limited compared to the underlying power it has. Moving it one layer up from the modules, allows the use of its logic by any module while maintaining each module functionality regarding specific DSN or security, for example, instead of introducing all that logic into the `sql` module.

The way it works now also allows to create metricsets on CockroachDB module using lightweight modules.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

Fixes: https://github.com/elastic/beats/issues/18898
Implements: https://github.com/elastic/beats/issues/15048

### CODE REVIEW

**Comment by sayden** [comment]
I'm thinking to remove the `overview` metricset to make this PR smaller and easier to review while we re-think what we want to include in the `overview` metricset or if we want to include it at all.

Having it as a lightweight module, it easy to include it later.

File: metricbeat/docs/modules/sql.asciidoc [hunk 1]
```diff
@@ -8,9 +8,361 @@ This file is generated! See scripts/mage/docs_collector.go
...
+
+Keys in the YAML are defined as follow:
+
+- `driver`: The drivers currently supported are those which already have a Metricbeat module like `mssql` or `postgres`.
+- `sql_query`: Is the single query you want to run
+- `sql_response_format`: You have 2 options here:
+    - `variables`: Expects a table which looks like a key/value result. With 2 columns, left column will be considered a key and the right column the value. This mode generates a single event on each fetch operation.
+    - `table`: Table mode can contain any number of columns and a single event will be generated for each row.
+
+Results will be grouped by type in the result event for convenient mapping in Elasticsearch. So `strings` values will be grouped into `sql.strings`, `numeric` into `sql.numeric` and so on and so for.
```
Comment (ChrsMark) [inline metricbeat/docs/modules/sql.asciidoc:61]:
```suggestion
Results will be grouped by type in the result event for convenient mapping in Elasticsearch. So `strings` values will be grouped into `sql.strings`, `numeric` into `sql.numeric` and so on and so forth.
```
;)

File: metricbeat/docs/modules/sql.asciidoc [hunk 2]
```diff
@@ -8,9 +8,361 @@ This file is generated! See scripts/mage/docs_collector.go
...
+  },
+  "host": {
+    "name": "elastic"
+  }
+}
+----
+
+=== Two or more queries
+
+If you want to launch two or more queries, you need to specify them with their full configuration for each query. For example:
```
Comment (ChrsMark) [inline metricbeat/docs/modules/sql.asciidoc:343]: Not sure if it would be more convenient to do something similar to Prometheus Query metricset where we define multiple queries within a single module config. see https://github.com/elastic/beats/blob/master/metricbeat/modules.d/prometheus.yml.disabled#L37

File: metricbeat/helper/sql/sql.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	"database/sql"
+	"fmt"
+
+	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/lib/pq"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/pkg/errors"
+	"strconv"
```
Comment (ChrsMark) [inline metricbeat/helper/sql/sql.go:14]: nit: import order :)

File: metricbeat/helper/sql/sql.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	"strings"
+	"time"
+)
+
+type Db struct {
```
Comment (ChrsMark) [inline metricbeat/helper/sql/sql.go:19]: `DbClient` is maybe a more self-explanatory name?
Also does this struct need to be exported? Could it be just `dbClient` ?

File: metricbeat/helper/sql/sql.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+type sqlRow interface {
+	Scan(dest ...interface{}) error
+	Next() bool
+	Columns() ([]string, error)
+	Err() error
+}
+
+// DB gets a client ready to query the database
+func DB(driver, uri string, l *logp.Logger) (*Db, error) {
```
Comment (ChrsMark) [inline metricbeat/helper/sql/sql.go:32]: Not strong opinion but maybe `CreateDBClient` could be more handy for future code visitors.

File: metricbeat/helper/sql/sql.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+// 'godror' so this detail implementation must be hidden to the user, that should only choose and see 'oracle' as driver
+func switchDriverName(d string) string {
+	switch d {
+	case "oracle":
+		return "godror"
+	case "cockroachdb":
+		return "postgres"
+	case "cockroach":
+		return "postgres"
+	case "postgresql":
```
Comment (ChrsMark) [inline metricbeat/helper/sql/sql.go:182]: `mysql` driver?

File: x-pack/metricbeat/module/sql/query/query.go [hunk 7]
```diff
@@ -7,14 +7,13 @@ package query
 import (
 	"context"
 	"fmt"
-	"strconv"
-	"strings"
-	"time"
+
+	"github.com/elastic/beats/v7/libbeat/common"
```
Comment (ChrsMark) [inline x-pack/metricbeat/module/sql/query/query.go:11]: nit: import order

File: metricbeat/module/mysql/overview/overview.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+http://dev.mysql.com/doc/refman/5.7/en/show-status.html
+*/
+package overview
+
+import (
+	"context"
+	"github.com/elastic/beats/v7/metricbeat/helper/sql"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/metricbeat/module/mysql"
+	"github.com/pkg/errors"
```
Comment (ChrsMark) [inline metricbeat/module/mysql/overview/overview.go:31]: nit: import order

File: metricbeat/module/mysql/overview/manifest.yml [hunk 9]
```diff
@@ -0,0 +1,25 @@
+default: true
+input:
+  module: mysql
+  metricset: query
+  defaults:
+    sql_response_format: variable
+    namespace: overview
+    sql_queries:
+      - query: >
+          SELECT *
```
Comment (ChrsMark) [inline metricbeat/module/mysql/overview/manifest.yml:10]: 🤔  I'm a little bit confused :) , is this a light module? Cause I see the same query in implementation in `metricbeat/module/mysql/overview/overview.go`.

File: metricbeat/helper/sql/sql_test.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
+package sql
+
+import (
+	"github.com/elastic/beats/v7/libbeat/common"
```
Comment (ChrsMark) [inline metricbeat/helper/sql/sql_test.go:4]: nit: import order

File: metricbeat/helper/sql/sql.go [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+	for key, value := range data {
+		value := getValue(&value)
+		r.Put(key, value)
+	}
+
+	return r, nil
+}
+
+func ToDotKeys(ms common.MapStr) common.MapStr {
```
Comment (ChrsMark) [inline metricbeat/helper/sql/sql.go:138]: docstring?

File: metricbeat/docs/modules/sql.asciidoc (hunk 2 above)
Comment (sayden) [reply metricbeat/docs/modules/sql.asciidoc:343]: I see your point and, in fact, it is implemented that way in the MySQL module (also in this PR) but I'd prefer to not add more complexity to the PR (it's already very big) and leave that change for a future discussion. The reason is because I feel like it's not a very good idea to make the `sql` module very big and / or "intelligent" and rather give more power to `query` metricsets in the rest of the modules, which can handle more things like the DSNs, auth, etc.

How does it sounds?

File: metricbeat/helper/sql/sql.go (hunk 4 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go:19]: Yeah, I think I'll change the name. This code comes directly from the `sql` module and I wanted to maintain as many things as possible like the original code. But if two people (you and me) think that it may have a better name... I didn't change it before because this code has passed already a review without complains about it :smile:

File: metricbeat/helper/sql/sql.go (hunk 5 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go:32]: The reason I didn't call it `GetDb` or `CreateDb` is this https://golang.org/doc/effective_go.html#Getters

File: metricbeat/helper/sql/sql.go (hunk 6 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go:182]: `mysql` do not require a "translation" as everybody call it `mysql`. Example: The fact that the library of oracle is called `godror` is an implementation detail that Beats shouldn't expose to users. Users will expect to write `oracle` as the driver name. Same with CockroachDB which uses `postgres` driver or `postgresql ~= postgres`.

It can be very annoying to write `postgresql` as the driver name and receive a `driver not recognized` because the actual driver name is `postgres` (without `ql`)

File: metricbeat/helper/sql/sql.go (hunk 11 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go:138]: docstring? I don't follow you :smile:

File: metricbeat/module/mysql/overview/manifest.yml (hunk 9 above)
Comment (sayden) [reply metricbeat/module/mysql/overview/manifest.yml:10]: Yeah, I initially created it as a "normal" module and then I realized that I could try to invest some time in doing it as a lightweight module. I forgot completely about this file. I have just deleted it.
Thanks for the heads up :smiley:

File: metricbeat/helper/sql/sql.go (hunk 4 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go:19]: Ah! BTW, it must be exported as it's used by every Metricset which takes advantage of it

File: metricbeat/module/mysql/query/query.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+		if q.QueryNamespace != "" {
+			event.ModuleFields[m.Config.Namespace] = common.MapStr{q.QueryNamespace: sql.ToDotKeys(ms)}
+		} else {
+			event.ModuleFields[m.Config.Namespace] = sql.ToDotKeys(ms)
+		}
+		reporter.Event(event)
+	}
+
+	return nil
+}
```
Comment (jsoriano) [inline metricbeat/module/mysql/query/query.go:124]: For a future change, the namespace logic might go to the helper, I guess other metricsets may benefit of this.

File: x-pack/metricbeat/module/sql/query/query.go [hunk 13]
```diff
@@ -81,167 +79,81 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
...
-		num, err := strconv.ParseFloat(value, 64)
-		if err == nil {
-			numericMetrics[key] = num
-		} else {
-			stringMetrics[key] = value
+	boolMetrics := common.MapStr{}
+
+	for k, v := range ms {
+		switch v.(type) {
+		case float64:
```
Comment (jsoriano) [inline x-pack/metricbeat/module/sql/query/query.go:131]: Are all numeric values typed as `float64`?

**Comment by sayden** [comment]
Very good comments indeed. Thanks! Will take a look

File: metricbeat/module/mysql/query/_meta/fields.yml [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+              description: >
+                The maximum length of the send queue since the last FLUSH STATUS command.
+
+            - name: queue_min
+              type: long
+              description: >
+                The minimum length of the send queue since the last FLUSH STATUS command.
+
+        - name: state
+          type: keyword
```
Comment (kaiyan-sheng) [inline metricbeat/module/mysql/query/_meta/fields.yml:217]: I don't know what the value of this state look like. Should this state number be int or keyword?

File: metricbeat/module/mysql/query/query.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+		m.db, err = sql.DB("mysql", m.HostData().URI, m.Logger())
+		if err != nil {
+			return errors.Wrap(err, "mysql-status fetch failed")
+		}
+	}
+
+	for _, q := range m.Config.Queries {
+		err := m.fetchQuery(ctx, q, reporter)
+		if err != nil {
+			m.Logger().Error("error doing query", err)
```
Comment (kaiyan-sheng) [inline metricbeat/module/mysql/query/query.go:85]: Would be helpful to add the query into the error message?

File: metricbeat/docs/modules/sql.asciidoc (hunk 2 above)
Comment (ChrsMark) [reply metricbeat/docs/modules/sql.asciidoc:343]: Fair enough! Thanks for elaborating!

File: metricbeat/helper/sql/sql.go (hunk 4 above)
Comment (ChrsMark) [reply metricbeat/helper/sql/sql.go]: 👍

File: metricbeat/helper/sql/sql.go (hunk 5 above)
Comment (ChrsMark) [reply metricbeat/helper/sql/sql.go]: :+1 for not having a `Get` prefix but how about `DBClient`?

File: metricbeat/helper/sql/sql.go (hunk 6 above)
Comment (ChrsMark) [reply metricbeat/helper/sql/sql.go:200]: Gotcha, thanks!

File: metricbeat/module/mysql/overview/manifest.yml (hunk 9 above)
Comment (ChrsMark) [reply metricbeat/module/mysql/overview/manifest.yml]: 👍

File: metricbeat/helper/sql/sql.go [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+type sqlRow interface {
+	Scan(dest ...interface{}) error
+	Next() bool
+	Columns() ([]string, error)
+	Err() error
+}
+
+// DB gets a client ready to query the database
+func DB(driver, uri string, l *logp.Logger) (*DbClient, error) {
```
Comment (ChrsMark) [inline metricbeat/helper/sql/sql.go:51]: Since we changed the struct's name, should we also tune the getter according to https://golang.org/doc/effective_go.html#Getters? :)

Comment (jsoriano) [reply metricbeat/helper/sql/sql.go:51]: Actually this is more a constructor than a getter, isn't it? https://golang.org/doc/effective_go.html#composite_literals :slightly_smiling_face: :slightly_smiling_face:

Comment (sayden) [reply metricbeat/helper/sql/sql.go:51]: Yeah, worth mentioning that even small changes in code or naming can cause chain-reactions in many other places and this is a good example of it

File: metricbeat/module/mysql/query/query.go (hunk 15 above)
Comment (sayden) [reply metricbeat/module/mysql/query/query.go:85]: The query for MySQL module will be static so I'm not sure if it may lead to confusion. However in the `sql` module the query is an input of the user so I'll definitely add it there.

File: metricbeat/docs/modules/sql.asciidoc [hunk 17]
```diff
@@ -8,9 +8,361 @@ This file is generated! See scripts/mage/docs_collector.go
 
 beta[]
 
-This is the sql module that fetches metrics from a SQL database. You can define driver and SQL query.
+The SQL module allows to fetch custom queries from a SQL database to send them to Elasticsearch.
```
[the remark this answers is not in the document]
Comment (sayden) [reply metricbeat/docs/modules/sql.asciidoc]: I have done the suggestion. Thanks!

File: metricbeat/docs/modules/sql.asciidoc (hunk 1 above)
Comment (sayden) [reply metricbeat/docs/modules/sql.asciidoc]: I have fixed the typo. Thanks!

File: metricbeat/helper/sql/sql.go (hunk 3 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go:28]: Reordered

File: metricbeat/helper/sql/sql.go (hunk 5 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go]: If the type is called `DbClient` at the package level, function cannot be called `DbClient` at the same package level :smile: So probably `NewDbClient` is the best candidate

File: x-pack/metricbeat/module/sql/query/query.go (hunk 7 above)
Comment (sayden) [reply x-pack/metricbeat/module/sql/query/query.go]: Reordered :slightly_smiling_face:

File: metricbeat/module/mysql/overview/overview.go (hunk 8 above)
Comment (sayden) [reply metricbeat/module/mysql/overview/overview.go]: I finally removed the `overview` metricset for a follow up PR. Thanks!

File: metricbeat/helper/sql/sql.go (hunk 11 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go:156]: Gotcha! :smile:

File: metricbeat/helper/sql/sql.go (hunk 16 above)
Comment (ChrsMark) [reply metricbeat/helper/sql/sql.go:51]: Isn't refactoring/renaming of functions/variables handled by editors quite smoothly? In any case if it's about to cause other issues we can keep it as is!

File: metricbeat/module/mysql/query/query.go [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet("mysql", "query", New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+type query struct {
+	QueryNamespace string `config:"query_namespace"`
```
[the remark this answers is not in the document]
Comment (sayden) [reply metricbeat/module/mysql/query/query.go:44]: Yeah, it's to avoid confusion with the metricset namespace. I got confused a couple of times myself while testing the module and I thought it was a bit too implicit to see. What I can do is to change the variable name inside the code but I think that the user facing yaml should remain as `query_namespace`.

Actually, I wanted to call the other `metricset_namespace` too :smile: The issue is that `namespace` is so widely used everywhere that it has way too many meanings.

[the remark this answers is not in the document]
Comment (sayden) [reply metricbeat/module/mysql/query/query.go:44]: This is how it looks now: https://github.com/elastic/beats/pull/18955/files#diff-4315d136c22767eae25860e395f2798cR7

File: CHANGELOG.next.asciidoc [hunk 19]
```diff
@@ -533,6 +533,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Add support for named ports in autodiscover. {pull}19398[19398]
 - Add param `aws_partition` to support aws-cn, aws-us-gov regions. {issue}18850[18850] {pull}19423[19423]
 - The `elasticsearch/index` metricset now collects metrics for hidden indices as well. {issue}18639[18639] {pull}18703[18703]
+- Added to MySQL `performance` as lightweight metricset and `query` as parent lightweight metricset {pull}18955[18955]
```
Comment (jsoriano) [inline CHANGELOG.next.asciidoc:542]: Users don't mind if they are lightweight or not :slightly_smiling_face: 
```suggestion
- Added `performance` and `query` metricsets to `mysql` module. {pull}18955[18955]
```

File: metricbeat/module/mysql/performance/_meta/fields.yml [hunk 20]
```diff
@@ -1,6 +1,49 @@
...
+      fields:
+        - name: 'max.timer.wait'
+          type: long
+          description: Maximum wait time of the summarized events that are timed
+        - name: 'last.seen'
+          type: date
+          description: Time at which the digest was most recently seen
+        - name: 'quantile.95'
+          type: long
+          description: TODO
```
Comment (jsoriano) [inline metricbeat/module/mysql/performance/_meta/fields.yml:19]: Add a description here before merging?

File: metricbeat/module/mysql/query/query.go [hunk 21]
```diff
@@ -89,40 +89,47 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
...
 	}
 
 	return nil
 }
 
+func (m *MetricSet) transformMapStrToEvent(query query, ms common.MapStr) mb.Event {
+	event := mb.Event{ModuleFields: common.MapStr{m.Config.Namespace: common.MapStr{}}}
+
+	data := ms
+	if query.DeDotEnabled {
```
Comment (jsoriano) [inline metricbeat/module/mysql/query/query.go:120]: Would it be the oposite? When `!query.DeDotEnabled`, then we add the dots.

File: metricbeat/module/mysql/query/query.go [hunk 22]
```diff
@@ -41,18 +40,19 @@ func init() {
 }
 
 type query struct {
-	QueryNamespace string `config:"query_namespace"`
-	Query          string `config:"query"`
+	Namespace      string `config:"query_namespace" validate:"nonzero,required"`
```
Comment (jsoriano) [inline metricbeat/module/mysql/query/query.go:43]: The problem of forcing the use of a custom namespace in a generic module is that we cannot provide default mappings as we do in the sql module with the `sql.metrics...` fields.
We should probably use some namespace by default, and provide a mapping for it.

File: x-pack/metricbeat/module/sql/query/query.go [hunk 23]
```diff
@@ -81,167 +79,81 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
...
+			numericMetrics.Put(k, v)
+		case string:
+			stringMetrics.Put(k, v)
+		case bool:
+			boolMetrics.Put(k, v)
+		case nil:
+		//Ignore because a nil has no data type and thus cannot be indexed
```
Comment (jsoriano) [inline x-pack/metricbeat/module/sql/query/query.go:138]: Changes here would also require a changelog entry, or to be moved to a different PR.

File: metricbeat/module/mysql/query/query.go [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+	mb.Registry.MustAddMetricSet("mysql", "query", New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+type query struct {
+	Namespace      string `config:"query_namespace" validate:"nonzero,required"`
+	Query          string `config:"query" validate:"nonzero,required"`
+	ResponseFormat string `config:"response_format" validate:"nonzero,required"`
+	DeDotEnabled   bool   `config:"dedot.enabled"`
```
Comment (jsoriano) [inline metricbeat/module/mysql/query/query.go:46]: We are dedotting by default everywhere else, we should set it to true by default implementing `InitDefaults()`.
Also I am not sure if we should use the `dedot` name here, as we are using it to do the oposite, adding dots where there weren't before.

File: metricbeat/module/mysql/query/query.go [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+// MetricSet for fetching MySQL server status.
+type MetricSet struct {
+	mb.BaseMetricSet
+	db     *sql.DbClient
+	Config struct {
+		Queries   []query `config:"queries" validate:"nonzero,required"`
+		Namespace string  `config:"namespace" validate:"nonzero,required"`
+	}
```
Comment (jsoriano) [inline metricbeat/module/mysql/query/query.go:56]: This metricset would need documentation. I am ok with adding it in a follow up PR in any case.

File: metricbeat/module/mysql/performance/manifest.yml [hunk 26]
```diff
@@ -0,0 +1,22 @@
+default: true
+input:
+  module: mysql
+  metricset: query
+  defaults:
+    namespace: performance
+    queries:
+      - query: >
+          SELECT digest_text, count_star, avg_timer_wait, max_timer_wait, last_seen, quantile_95
```
Comment (jsoriano) [inline metricbeat/module/mysql/performance/manifest.yml:9]: Btw, would be an alternative to adding dots, to explicitly put them already here?
```suggestion
          SELECT digest_text AS digest.text, count_star AS count.star, avg_timer_wait AS avg.timer.wait, max_timer_wait AS max.timer.wait, last_seen AS last.seen, quantile_95 AS quantile.95
```

If this work, this would give more control on the field names, and we wouldn't need an additional option to add dots.

Comment (jsoriano) [reply metricbeat/module/mysql/performance/manifest.yml:9]: To use dots, names need to be quoted. I have updated the previous comment to use quotes in the suggestion.

Comment (sayden) [reply metricbeat/module/mysql/performance/manifest.yml:9]: Yeap, this works in table response mode where you can choose the columns names. However it doesn't in variable mode where field name is a value in a `name` column. I introduced the `dedot.enabled` field to activate or deactivate it anyways.

File: metricbeat/module/mysql/query/query.go (hunk 25 above)
Comment (sayden) [reply metricbeat/module/mysql/query/query.go:57]: I'd like to write it now, but we need to confirm the behavior and naming :smile:

File: metricbeat/module/mysql/query/query.go (hunk 24 above)
Comment (sayden) [reply metricbeat/module/mysql/query/query.go:47]: Good catch, I took the name from `http.json` metricset but I read its behavior in reverse. Any suggestions for the name?

File: metricbeat/helper/sql/sql.go (hunk 16 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go:51]: Oh yes, I'd love that IDE's take care of **proper naming** (not renaming / refactoring) for us :smile: But I think that AI hasn't reached that level yet :wink:

File: x-pack/metricbeat/module/sql/query/query.go (hunk 13 above)
Comment (sayden) [reply x-pack/metricbeat/module/sql/query/query.go:131]: I forgot to answer this. I spent quite a lot of time with this: The issue is that sql drivers handle numeric type inconsistently between them and the generic result is `float64`. I mean that while driver A handles `INTEGER` sql type properly and returns `int32` go types, driver B handles all `INTEGER` sql type as `float64`. Same problem with `NUMERIC`.

Sad story here is that Go's SQL driver has the tools to deal with this properly https://golang.org/src/database/sql/driver/driver.go?s=15846:15947#L413 but not all drivers implement it

Comment (jsoriano) [reply x-pack/metricbeat/module/sql/query/query.go:131]: So is `float64` enough here or we should add some other numeric type to be covered?

File: metricbeat/module/mysql/query/_meta/fields.yml (hunk 14 above)
Comment (sayden) [reply metricbeat/module/mysql/query/_meta/fields.yml]: This has been removed and it will be included in a follow up PR. I... frankly don't remember now what was that value anyways :smile:

File: metricbeat/helper/sql/sql.go [hunk 27]
```diff
@@ -0,0 +1,205 @@
...
+
+package sql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/lib/pq"
```
[the remark this answers is not in the document]
Comment (sayden) [reply metricbeat/helper/sql/sql.go]: Removed

File: x-pack/metricbeat/module/sql/query/query.go (hunk 13 above)
Comment (sayden) [reply x-pack/metricbeat/module/sql/query/query.go:131]: Support for specific numeric types is out of the scope of this PR and it's not supported by the original module either. We can open an issue later to discuss about complexity and benefits first.

File: metricbeat/helper/sql/sql.go (hunk 16 above)
Comment (sayden) [reply metricbeat/helper/sql/sql.go]: I left it with `type DbClient` and `func NewDbClient`

File: CHANGELOG.next.asciidoc (hunk 19 above)
Comment (sayden) [reply CHANGELOG.next.asciidoc:542]: Ok, changed.

File: metricbeat/module/mysql/query/query.go (hunk 21 above)
Comment (sayden) [reply metricbeat/module/mysql/query/query.go]: It was finally replaced with `ReplaceUnderscores`

File: metricbeat/module/mysql/query/query.go (hunk 22 above)
Comment (sayden) [reply metricbeat/module/mysql/query/query.go]: I'll open an issue to add default mappings to `query` metricset because they can use `sql` module as a workaround. This way I don't make this PR even bigger.

File: x-pack/metricbeat/module/sql/query/query.go (hunk 23 above)
Comment (sayden) [reply x-pack/metricbeat/module/sql/query/query.go:138]: Added

File: metricbeat/module/mysql/performance/_meta/data.json [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+    },
+    "agent": {
+        "name": "mcastro",
+        "type": "metricbeat",
+        "version": "8.0.0",
+        "ephemeral_id": "c1f810aa-6f0e-4550-b29e-96ee1f402d65",
+        "id": "de510575-3657-4099-812a-aafd2febc68e"
+    },
+    "service": {
+        "address": "root:root@tcp(172.18.0.2:3306)/",
```
[the remark this answers is not in the document]
Comment (sayden) [reply metricbeat/module/mysql/performance/_meta/data.json:19]: There's a bug with the HostParser in Lightweight modules which blocks this PR. The HostParser is being executed twice, once with the original incoming config and once with the "resulting" config. We should avoid the double execution.

File: metricbeat/module/mysql/performance/manifest.yml [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+        query_namespace: events_statements
+        response_format: table
+        dedot.enabled: false
+      - query: >
+          SELECT object_schema, object_name, index_name, count_fetch
+          FROM performance_schema.table_io_waits_summary_by_index_usage
+          WHERE count_fetch > 0
+        query_namespace: table_io_waits
+        response_format: table
+        dedot.enabled: false
```
[the remark this answers is not in the document]
Comment (sayden) [reply metricbeat/module/mysql/performance/manifest.yml:22]: Changed

File: metricbeat/module/mysql/query/query.go [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/v7/metricbeat/helper/sql"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet("mysql", "query", New,
+		mb.DefaultMetricSet(),
+	)
```
[the remark this answers is not in the document]
Comment (sayden) [reply metricbeat/module/mysql/query/query.go:40]: See this https://github.com/elastic/beats/pull/18955#discussion_r453564818

File: metricbeat/module/mysql/performance/_meta/fields.yml (hunk 20 above)
Comment (sayden) [reply metricbeat/module/mysql/performance/_meta/fields.yml]: Done!

File: metricbeat/docs/modules/sql.asciidoc [hunk 31]
```diff
@@ -8,9 +8,361 @@ This file is generated! See scripts/mage/docs_collector.go
...
+
+- module: sql
+  metricsets:
+    - query
+  period: 10s
+  hosts: ["postgres://postgres:postgres@localhost:5432/stuff?sslmode=disable"]
+  driver: "postgres"
+  sql_query: "SELECT * FROM pg_catalog.pg_tables pt WHERE schemaname ='pg_catalog'"
+  sql_response_format: table
+---
```
Comment (bmorelli25) [inline metricbeat/docs/modules/sql.asciidoc:365]: This should fix the doc build
```suggestion
----
```

Comment (bmorelli25) [reply metricbeat/docs/modules/sql.asciidoc:365]: Oh but this is generated. I don't know where this is is coming from. @dedemorton probably will

Comment (sayden) [reply metricbeat/docs/modules/sql.asciidoc:365]: Oh please, I cannot believe it :smile:

Comment (sayden) [reply metricbeat/docs/modules/sql.asciidoc]: Ok, pushed and fixed. Thank you very much :smiley:
